### ISSUE TITLE
Extend the experimental comparison between the parallelization strategies for AND, XOR and OR queries to assess performance

### ISSUE THREAD

**Issue by lemire** [RoaringBitmap/RoaringBitmap#223]
We recently adopted parallelized functions for the aggregations of bitmaps, e.g.,  ``ParallelAggregation.or``. See https://github.com/RoaringBitmap/RoaringBitmap/pull/211

Optimization in such cases is a multidimensional problem, but at least for ``buffer`` bitmaps, it seems that there might be room for better or different trade-offs. However, we lack hard data at the moment.

What complicates matter is that different people might have different test cases. To keep things sane, we have an "official" set of real datasets: census-income, census1881, dimension_008,...

So it seems that we should benchmark the current parallelized implementations (e.g., ``ParallelAggregation.or``) against alternatives on our "standard" datasets to help gain more insights.

Further documentation:

User @ppiotrow writes:

> I've compared Maciej's and this implementation in our production application for two very heavy queries. This is only about **buffered** bitmaps.

```
Maciej’s (parr.OR + parr.AND)
[info] Benchmark                                Mode  Cnt     Score     Error  Units
[info] ProdApp.benchA                   	avgt   20  2873.975 ± 574.882  ms/op
[info] ProdApp.benchB                    	avgt   20   728.568 ±  70.563  ms/op

Maciej's (parr.OR)
[info] Benchmark                                Mode  Cnt     Score     Error  Units
[info] ProdApp.benchA                   	avgt   20  5652.284 ± 765.898  ms/op
[info] ProdApp.benchB                    	avgt   20  1550.732 ± 654.949  ms/op

Official
[info] Benchmark                                Mode  Cnt     Score     Error  Units
[info] ProdApp.benchA                   	avgt   20  6351.598 ± 508.392  ms/op
[info] ProdApp.benchB                    	avgt   20  2915.789 ± 571.573  ms/op
```
> Two conclusions, the parallel `AND` makes difference and in general @maciej approach was faster even with only `OR` enabled. Sorry, but I cannot provide benchmark data or queries to you, but those are in general alternatives and conjunctions of about 3-100 bitmaps which has values around from 0 to 500 000 000.

User @richardstartin wrote:

> As has been alluded to throughout this PR, there are a few dimensions to this problem (number of bitmaps, size of bitmaps, overlap of keys). An alternative implementation was previously welcomed to be benchmarked against this implementation, on the real datasets, and that offer is obviously still open.

**Comment by richardstartin** [comment]
Some big improvements weren't ported over to the buffer package before merge, which especially apply to skewed aggregations. This is a case of copy and paste, which I will issue another PR for, not for it necessarily to be merged, but for @ppiotrow to run his private benchmark against. 

I'd also caution aiming for one "best" implementation.

**Comment by lemire** [comment]
> I'd also caution aiming for one "best" implementation.

Right. But sometimes, when we understand better the trade-offs, we can offer different strategies. It is possible to offer more than just one function for each task.

**Comment by richardstartin** [comment]
I agree.

Here are the key overlaps for the data sets used for benchmarking currently. Each number is how many bitmaps in the aggregation have a container at a given key. 

```
census-income: [196, 196, 195, 181]
census1881: [23, 27, 23, 28, 20, 23, 20, 20, 19, 26, 26, 23, 21, 21, 14, 17, 20, 20, 19, 21, 21, 18, 23, 24, 18, 24, 19, 20, 17, 23, 26, 27, 19, 24, 25, 22, 22, 23, 24, 19, 21, 19, 21, 19, 26, 24, 22, 20, 17, 20, 29, 22, 26, 24, 23, 24, 22, 29, 23, 23, 22, 25, 24, 23, 25, 22]
dimension_008: [4444, 799, 995, 15, 100, 69, 18, 58, 197, 199, 4, 17, 58, 66, 4, 7, 7, 11, 53, 23, 42, 38, 45, 106, 24, 88, 88, 3, 2, 3, 1, 5, 5, 6, 5, 11, 58, 55, 46, 54, 648, 316, 17, 10, 6, 5, 1, 3, 8, 12, 19, 8, 1, 95, 111, 5]
dimension_003: [1996, 1560, 1928, 128, 138, 322, 160, 206, 515, 525, 121, 89, 190, 253, 92, 117, 117, 150, 159, 97, 110, 76, 77, 152, 54, 242, 276, 3, 14, 4, 81, 76, 39, 10, 32, 10, 23, 15, 20, 29, 392, 893, 298, 313, 1721, 838, 669, 91, 35, 60, 9, 28, 30, 62, 64, 71, 3, 664, 249, 9]
dimension_033: [62, 42, 64, 43, 28, 52, 31, 30, 46, 65, 42, 24, 38, 57, 44, 41, 45, 47, 32, 29, 31, 22, 19, 33, 23, 61, 66, 3, 5, 3, 28, 31, 24, 8, 12, 8, 14, 8, 13, 15, 45, 30, 23, 18, 29, 43, 59, 26, 11, 21, 4, 12, 12, 18, 14, 17, 3, 55, 32, 7]
uscensus2000: [6, 6, 6, 6, 4, 10, 13, 10, 2, 3, 5, 4, 3, 9, 10, 5, 4, 3, 5, 5, 5, 9, 5, 1, 1, 4, 4, 4, 6, 3, 5, 2, 4, 5, 3, 9, 5, 7, 1, 2, 4, 5, 6, 7, 14, 17, 11, 11, 14, 10, 20, 25, 2, 3, 2, 5, 2, 5, 6, 6, 4, 3, 4, 4, 2, 7, 8, 1, 1, 1, 3, 3, 2, 3, 4, 4, 4, 4, 3, 4, 7, 6, 1, 1, 1, 3, 1, 3, 4, 3, 3, 7, 6, 4, 8, 11, 4, 1, 1, 3, 1, 2, 3, 5, 1, 3, 3, 2, 7, 8, 4, 1, 2, 2, 1, 3, 3, 1, 1, 1, 4, 1, 3, 4, 3, 1, 2, 2, 1, 5, 4, 2, 6, 2, 7, 3, 4, 9, 5, 5, 4, 5, 3, 5, 7, 6, 1, 4, 3, 5, 3, 4, 6, 3, 2, 2, 3, 2, 6, 1, 1, 1, 3, 4, 2, 4, 6, 3, 2, 2, 5, 2, 4, 6, 4, 3, 2, 3, 4, 3, 7, 6, 1, 2, 2, 4, 4, 3, 6, 5, 1, 3, 3, 4, 4, 8, 6, 2, 2, 3, 5, 3, 5, 7, 3, 1, 3, 4, 3, 5, 8, 8, 3, 3, 5, 6, 4, 6, 5, 2, 4, 3, 4, 3, 4, 5, 3, 1, 3, 4, 3, 1, 6, 5, 5, 3, 5, 8, 3, 9, 5, 6, 2, 5, 3, 2, 4, 4, 3, 2, 5, 6, 4, 4, 10, 5, 1, 3, 2, 2, 5, 2, 1, 1, 1, 3, 1, 4, 2, 2, 1, 1, 2, 2, 4, 4, 2, 2, 3, 3, 3, 3, 6, 5, 3, 2, 3, 5, 3, 4, 6, 5, 2, 4, 6, 4, 3, 6, 7, 1, 1, 3, 2, 3, 3, 3, 1, 4, 3, 3, 6, 3, 1, 1, 3, 1, 2, 3, 1, 2, 2, 1, 3, 3, 2, 1, 2, 2, 1, 3, 2, 2, 2, 2, 5, 1, 4, 5, 2, 1, 2, 3, 3, 2, 4, 3, 1, 1, 2, 3, 1, 3, 4, 1, 1, 2, 3, 2, 5, 2, 2, 2, 5, 2, 3, 5, 5, 5, 6, 11, 10, 13, 9, 8, 2, 2, 3, 1, 4, 3, 1, 1, 1, 4, 1, 3, 4, 2, 1, 2, 2, 1, 4, 5, 6, 3, 4, 8, 6, 7, 9, 9, 4, 4, 6, 6, 9, 11, 9, 3, 3, 5, 9, 6, 11, 10, 4, 2, 3, 3, 4, 5, 5, 4, 1, 3, 1, 3, 2, 6, 3, 1, 1, 1, 3, 1, 2, 3, 7, 4, 4, 7, 7, 6, 20, 12, 3, 5, 3, 7, 6, 9, 8, 5, 2, 2, 3, 5, 4, 7, 6, 2, 2, 2, 5, 3, 4, 5, 3, 2, 2, 4, 3, 5, 5, 4, 2, 3, 2, 6, 3, 6, 6, 13, 11, 14, 15, 11, 21, 16, 15, 1, 3, 3, 2, 7, 2, 3, 2, 2, 6, 3, 4, 5, 4, 1, 1, 4, 2, 2, 6, 2, 1, 1, 3, 1, 2, 3, 2, 1, 3, 3, 3, 3, 4, 5, 1, 4, 3, 4, 3, 5, 4, 3, 1, 1, 2, 1, 3, 4, 2, 1, 1, 2, 2, 1, 5, 2]
weather_sept_85: [187, 182, 177, 176, 177, 175, 177, 174, 175, 176, 176, 176, 178, 184, 183, 183]
wikileaks-noquotes: [94, 98, 95, 97, 96, 95, 92, 94, 91, 95, 91, 90, 90, 85, 88, 91, 84, 79, 81, 73, 93]
census-income_srt: [193, 194, 188, 116]
census1881_srt: [36, 33, 33, 32, 33, 33, 39, 30, 40, 37, 31, 39, 31, 35, 35, 33, 37, 34, 36, 31, 28, 29, 34, 34, 37, 42, 30, 40, 44, 52, 47, 44, 30, 35, 41, 46, 40, 51, 45, 50, 44, 46, 29, 46, 46, 40, 44, 41, 45, 28, 39, 45, 47, 43, 42, 32, 46, 43, 42, 31, 46, 49, 34, 48, 29, 26]
weather_sept_85_srt: [54, 133, 164, 149, 108, 90, 131, 141, 151, 166, 161, 159, 138, 161, 139, 123]
wikileaks-noquotes_srt: [81, 93, 67, 76, 67, 70, 72, 68, 89, 65, 72, 76, 76, 80, 71, 85, 73, 73, 76, 84, 61]
```

This can be calculated with the following Java:

```java
int[] overlaps = ParallelAggregation.groupByKey(bitmaps).values().mapToInt(List::size).toArray();
```

For the sake of my curiosity, I would be fascinated to know these statistics from the aforementioned benchmark. Knowing this, and also the ratio of `RunContainer`, `BitmapContainer` and `ArrayContainer`s would allow for a simulation using the `RandomisedTestData` class.

**Comment by ppiotrow** [comment]
The statistics for both benchmarks' `or` queries.
https://gist.github.com/ppiotrow/b5597ec98e1b87964389e2c102bff442

How would you like me to print container types statistics?

**Comment by richardstartin** [comment]
Interesting, the aggregation is extremely sparse and narrow.

WRT container types, I don’t think it matters how it’s printed. Percentage per input bitmap? The point is, you can simulate your workload like this in about 20 mins without releasing data.

**Comment by richardstartin** [comment]
Am I right in thinking that a lot of the time you aren’t really aggregating containers, you’re really just appending them to an output? Some of those arrays are virtually all 1s with a handful of 2s and 3s. I think you could do this quickly on a single thread.

**Comment by ppiotrow** [comment]
Statistics of containers types per each OR operation
https://gist.github.com/ppiotrow/0cae260c61a49eba42df7a81cad2a9b4

Regarding 1s I think you're right. It would be faster to aggregate in the single thread. However as a simple RB user I don't have knowledge when to use parallel and when single threaded.

**Comment by richardstartin** [comment]
@ppiotrow Frankly, I think you're presenting a degenerate case. Nevertheless, it exists. Perhaps someone should think of a way to handle it. Please benchmark your implementation on data sets where there is a significant level of overlap and therefore parallelisble work.

**Comment by richardstartin** [comment]
@ppiotrow @maciej 

Let me know if the code at this [branch](https://github.com/maciej/RoaringBitmap/tree/par-poc) is the code you're benchmarking. If it is, have you tested that it produces the same output as `BufferFastAggregation` for a range of bitmaps? I just tried (renaming the class to avoid name clashes) and I don't think the code produces correct output: [see this test](https://github.com/richardstartin/RoaringBitmap/blob/par-poc/roaringbitmap/src/test/java/org/roaringbitmap/buffer/BufferFuzzer.java#L245). I'm tracking that branch now, so if there are any fixes to commit I can pick them up and benchmark them independently.

**Comment by ppiotrow** [comment]
Morning @richardstartin , it's clear that benchmark `A` is what we don't want to parallelise but would be great to detect it during calculations. Benchmark `B` however looks like the perfect scenario but it was slower than `Maciej's (parr.OR)` implementation.
Today I'll build the new improved code for buffered bitmaps and rerun the benchmarks and verify the cardinality for both approaches.

**Comment by richardstartin** [comment]
I fixed an unsigned error in the code on the branch and benchmarked against the "official" implementation on a 4 core/8 hyperthread machine. It's consistently a lot worse.


Benchmark | (dataset) | Mode | Cnt | Score | Error | Units | Ratio
-- | -- | -- | -- | -- | -- | -- | --
ParallelAggregatorBenchmark.bufferParallelOr | census-income | avgt | 5 | 340.713 | 232.272 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | census1881 | avgt | 5 | 496.425 | 3.137 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | dimension_008 | avgt | 5 | 773.783 | 21.557 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | dimension_003 | avgt | 5 | 1807.886 | 204.504 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | dimension_033 | avgt | 5 | 544.155 | 9.191 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | uscensus2000 | avgt | 5 | 179.714 | 3.062 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | weather_sept_85 | avgt | 5 | 1000.305 | 63.065 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | wikileaks-noquotes | avgt | 5 | 187.54 | 3.523 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | census-income_srt | avgt | 5 | 317.208 | 51.005 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | census1881_srt | avgt | 5 | 321.851 | 7.544 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | weather_sept_85_srt | avgt | 5 | 540.538 | 43.141 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOr | wikileaks-noquotes_srt | avgt | 5 | 125.579 | 1.398 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelOrMaciej | census-income | avgt | 5 | 953.023 | 80.233 | us/op | 2.797143
ParallelAggregatorBenchmark.bufferParallelOrMaciej | census1881 | avgt | 5 | 695.58 | 11.032 | us/op | 1.401178
ParallelAggregatorBenchmark.bufferParallelOrMaciej | dimension_008 | avgt | 5 | 3491.582 | 556.825 | us/op | 4.512353
ParallelAggregatorBenchmark.bufferParallelOrMaciej | dimension_003 | avgt | 5 | 7541.867 | 1710.984 | us/op | 4.17165
ParallelAggregatorBenchmark.bufferParallelOrMaciej | dimension_033 | avgt | 5 | 753.382 | 133.811 | us/op | 1.384499
ParallelAggregatorBenchmark.bufferParallelOrMaciej | uscensus2000 | avgt | 5 | 832.236 | 7.676 | us/op | 4.630891
ParallelAggregatorBenchmark.bufferParallelOrMaciej | weather_sept_85 | avgt | 5 | 2352.466 | 131.861 | us/op | 2.351749
ParallelAggregatorBenchmark.bufferParallelOrMaciej | wikileaks-noquotes | avgt | 5 | 568.942 | 4.669 | us/op | 3.03371
ParallelAggregatorBenchmark.bufferParallelOrMaciej | census-income_srt | avgt | 5 | 948.632 | 85.78 | us/op | 2.990568
ParallelAggregatorBenchmark.bufferParallelOrMaciej | census1881_srt | avgt | 5 | 561.282 | 7.92 | us/op | 1.743919
ParallelAggregatorBenchmark.bufferParallelOrMaciej | weather_sept_85_srt | avgt | 5 | 1474.426 | 30.659 | us/op | 2.727701
ParallelAggregatorBenchmark.bufferParallelOrMaciej | wikileaks-noquotes_srt | avgt | 5 | 305.352 | 5.621 | us/op | 2.431553

This isn't discounting your reported results, but it would be great to know what your actual benchmark is, and how many threads you're running. I would be interested to see your results of my benchmark on your benchmark machine. This is also to illustrate that this is a multifaceted issue.

Again, the branch is [here](https://github.com/richardstartin/RoaringBitmap/tree/par-poc).

**Comment by ppiotrow** [comment]
Looks great, I really hope that your implementation rules. I'll have time for that in few hour from now.

**Comment by richardstartin** [comment]
Sorry but I think you are personalising this issue. This isn’t about “my” code versus “maciej’s code“ - this is about the fastest implementation across a range of reproducible benchmark scenarios. You have reported an unverifiable result with a workload you can’t share or describe. If you could provide a public benchmark which reproduces the reported behaviour without revealing trade secrets, that would be a great contribution to the stability of the library.

**Comment by ppiotrow** [comment]
I didn't plan to make it personal, just a naming shortcut. I agree we are looking for the fastest solution. I'll benchmark it again on my machine soon. Regarding sharing the bitmaps I don't think it is possible.

**Comment by richardstartin** [comment]
I don't think it's a question of sharing the bitmaps, it's a question of writing the simplest benchmark scenario that can reproduce the issue, and sharing it for the sake of better benchmarking.

**Comment by ppiotrow** [comment]
After using the code from the results improved twice for benchmark `B` and both are at the same level as the compared implementation. For me it is green light to switch. We also have the same results in cardinality.
```
Benchmark                                Mode  Cnt     Score     Error  Units
[info] ProdApp.benchA                   avgt   20  5905.950 ± 652.147  ms/op
[info] ProdApp.benchB                   avgt   20  1653.192 ± 562.620  ms/op
```

What about parallel `And`? I think you mentioned that there is no performance gain in parallel approach.

**Comment by richardstartin** [comment]
The "container-parallel" code on that branch is level with master, it is forked from the "bitmap-parallel" [branch](https://github.com/maciej/RoaringBitmap/tree/par-poc) with a couple of bug fixes in the bitmap parallel implementation. 

Here are some benchmark results of the bitmap-parallel implementation. They are available to you to run. If you have some decent hardware, it would be interesting to see your results and compare. But on my machine, the parallel implementation is almost at parity with a single threaded implementation _once_ but is up to 122x worse. 


Benchmark | (dataset) | Mode | Cnt | Score | Error | Units | Ratio
-- | -- | -- | -- | -- | -- | -- | --
ParallelAggregatorBenchmark.bufferFastAnd | census-income | avgt | 5 | 8.306 | 1.108 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | census1881 | avgt | 5 | 3.389 | 0.213 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | dimension_008 | avgt | 5 | 121.291 | 18.533 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | dimension_003 | avgt | 5 | 431.684 | 248.316 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | dimension_033 | avgt | 5 | 153.695 | 110.294 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | uscensus2000 | avgt | 5 | 3.421 | 0.261 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | weather_sept_85 | avgt | 5 | 43.707 | 1.558 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | wikileaks-noquotes | avgt | 5 | 5.344 | 0.359 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | census-income_srt | avgt | 5 | 6.269 | 0.325 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | census1881_srt | avgt | 5 | 3.208 | 0.215 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | weather_sept_85_srt | avgt | 5 | 17.446 | 0.367 | us/op | 1
ParallelAggregatorBenchmark.bufferFastAnd | wikileaks-noquotes_srt | avgt | 5 | 6.856 | 7.674 | us/op | 1
ParallelAggregatorBenchmark.bufferParallelAndMaciej | census-income | avgt | 5 | 74.861 | 11.573 | us/op | 9.012882
ParallelAggregatorBenchmark.bufferParallelAndMaciej | census1881 | avgt | 5 | 172.373 | 2.966 | us/op | 50.8625
ParallelAggregatorBenchmark.bufferParallelAndMaciej | dimension_008 | avgt | 5 | 2280.536 | 204.925 | us/op | 18.80219
ParallelAggregatorBenchmark.bufferParallelAndMaciej | dimension_003 | avgt | 5 | 5894.186 | 385.855 | us/op | 13.65394
ParallelAggregatorBenchmark.bufferParallelAndMaciej | dimension_033 | avgt | 5 | 229.064 | 25.649 | us/op | 1.49038
ParallelAggregatorBenchmark.bufferParallelAndMaciej | uscensus2000 | avgt | 5 | 419.701 | 0.979 | us/op | 122.6837
ParallelAggregatorBenchmark.bufferParallelAndMaciej | weather_sept_85 | avgt | 5 | 301.501 | 49.629 | us/op | 6.898231
ParallelAggregatorBenchmark.bufferParallelAndMaciej | wikileaks-noquotes | avgt | 5 | 228.222 | 6.211 | us/op | 42.70621
ParallelAggregatorBenchmark.bufferParallelAndMaciej | census-income_srt | avgt | 5 | 75.157 | 21.106 | us/op | 11.98867
ParallelAggregatorBenchmark.bufferParallelAndMaciej | census1881_srt | avgt | 5 | 327.273 | 4.352 | us/op | 102.0178
ParallelAggregatorBenchmark.bufferParallelAndMaciej | weather_sept_85_srt | avgt | 5 | 253.25 | 2.872 | us/op | 14.51622
ParallelAggregatorBenchmark.bufferParallelAndMaciej | wikileaks-noquotes_srt | avgt | 5 | 207.206 | 1.343 | us/op | 30.22258

I think any of the following might add context to the discussion:

- a description of a use case
- a runnable benchmark
- `-prof perfasm`, `-prof gc`, _something_

I am stepping aside for now as I have nothing sensible to contribute.

**Comment by ppiotrow** [comment]
I confirm that the AND is indeed faster with single threaded benchmarks. I've run them on the benchmark machine from @lemire .

**Comment by ppiotrow** [comment]
The canary release of app with v. 0.7.8 shows very similar performance results as ones with comparable implementation. IMHO the issue #223 is resolved.

**Comment by richardstartin** [comment]
@ppiotrow that’s great news. Still, this has been at the back of my mind. If you have short fat aggregations (keys overlap, few keys per bitmap, lots of bitmaps) a bitmap-parallel implementation should be profitable for both AND and OR. If you have many overlapping keys, it’s terrible. I just don’t think there is a test dataset short and fat enough with enough overlapping keys to demonstrate this, though dimension_033 gets close. The client application would be the right place to store statistics about which parallel implementation to use, because detecting it here has a large cost.

**Comment by lemire** [comment]
Following @ppiotrow's comment, I am closing this issue.
